### PR TITLE
Lab test - Raspberry Pi OS arm64v8 build

### PR DESCRIPTION
**Pull request by frvfrvr**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.

## What's in this pull request?

> build_rpi64.Dockerfile for compatibility in Raspberry Pi OS arm64v8 (aarch64) 

**Device used**: Raspberry Pi 4 Model B Rev 1.4 (8GB RAM)
**Operating System installed**: `Linux raspberrypi 5.10.17-v8+ #1403 SMP PREEMPT Mon Feb 22 11:37:54 GMT 2021 aarch64 GNU/Linux` (Buster)
**Required** Latest 64bit beta OS image build from Raspberry Pi downloads: https://downloads.raspberrypi.org/raspios_arm64/images/

**Optional dependencies**: installing libseccomp2 with buster backports before performing docker build ([Source](https://blog.samcater.com/fix-workaround-rpi4-docker-libseccomp2-docker-20/)) 
```
# Get signing keys to verify the new packages, otherwise they will not install
rpi ~$ sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 04EE7237B7D453EC 648ACFD622F3D138

# Add the Buster backport repository to apt sources.list
rpi ~$ echo 'deb http://httpredir.debian.org/debian buster-backports main contrib non-free' | sudo tee -a /etc/apt/sources.list.d/debian-backports.list

rpi ~$ sudo apt update
rpi ~$ sudo apt install libseccomp2 -t buster-backports
```
Libseccomp2 has to be installed in case you encounter this error:
```
#8 ERROR: executor failed running [/bin/sh -c apt-get update && apt-get install -y libssl1.1]: exit code: 100
------
 > [stage-1 2/7] RUN apt-get update && apt-get install -y libssl1.1:
------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c apt-get update && apt-get install -y libssl1.1]: exit code: 100
```

Instead of `rustlang/rust:nightly-slim` as Rust image that is used to build Devlab, `arm64v8/rust:slim-buster` is used and installs a nightly version of it since slim-buster is stable release of Rust and doesn't work with compiling library parts that required nightly build.

### CODE REVIEW

**Comment by vlvrd** [comment]
@frvfrvr I'm testing on a Raspberry Pi 4 Model B Rev 1.2 with this branch, everything's updated, libseccomp2 is installed and I get this error:

```
#11 [builder 2/6] RUN rustup install nightly
#11 sha256:b6154b8f4a686c9c3f85e6f3457c2bdd3c8176f6ee2b86afcc35e32354c5ec2f
#11 0.960 standard_init_linux.go:219: exec user process caused: exec format error
#11 ERROR: executor failed running [/bin/sh -c rustup install nightly]: exit code: 1
------
 > [builder 2/6] RUN rustup install nightly:
------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c rustup install nightly]: exit code: 1
```

Seems as if I'm pulling something that's not meant for ARM. Has this happened to you? Am I missing a configuration step?

**Comment by frvfrvr** [comment]
> @frvfrvr I'm testing on a Raspberry Pi 4 Model B Rev 1.2 with this branch, everything's updated, libseccomp2 is installed and I get this error:
> 
> ```
> #11 [builder 2/6] RUN rustup install nightly
> #11 sha256:b6154b8f4a686c9c3f85e6f3457c2bdd3c8176f6ee2b86afcc35e32354c5ec2f
> #11 0.960 standard_init_linux.go:219: exec user process caused: exec format error
> #11 ERROR: executor failed running [/bin/sh -c rustup install nightly]: exit code: 1
> ------
>  > [builder 2/6] RUN rustup install nightly:
> ------
> error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c rustup install nightly]: exit code: 1
> ```
> 
> Seems as if I'm pulling something that's not meant for ARM. Has this happened to you? Am I missing a configuration step?

Thank you sir @vlvrd , may I know if the operating system shows aarch64 or arm64v8 and not armv71? in uname

tl;dr [Raspberry Pi OS 32-bit downloaded by default in Raspberry Pi website](https://www.raspberrypi.org/software/operating-systems/#raspberry-pi-os-32-bit) does not work with the packages installed by nimiq docker image, even if switching from `RUST_IMAGE=arm64v8` to `RUST_IMAGE=arm32v7` in build_aarch64.Dockerfile, this does not work in both 32-bit and 64-bit OS versions. 

Flashing a [Raspberry Pi OS 64-bit beta image](https://www.raspberrypi.org/forums/viewtopic.php?f=117&t=275370) will work.

> Linux raspberrypi 5.10.17-v8+ #1403 SMP PREEMPT Mon Feb 22 11:37:54 GMT 2021 **aarch64** GNU/Linux

-----
I've troubleshooted it with another Raspberry Pi 4 Model B Rev 1.2 (also 8GB) and running [Raspberry Pi OS 32-bit downloaded by default in Raspberry Pi website](https://www.raspberrypi.org/software/operating-systems/#raspberry-pi-os-32-bit).

```
#10 extracting sha256:0dbf61eeca21052d88ee62799e00af84d8b188451c0551fbd6819ff9356f1850 20.2s
#10 extracting sha256:0dbf61eeca21052d88ee62799e00af84d8b188451c0551fbd6819ff9356f1850 25.2s
#10 extracting sha256:0dbf61eeca21052d88ee62799e00af84d8b188451c0551fbd6819ff9356f1850 27.0s done
#10 DONE 125.9s

#11 [builder 2/6] RUN rustup install nightly
#11 sha256:b6154b8f4a686c9c3f85e6f3457c2bdd3c8176f6ee2b86afcc35e32354c5ec2f
#11 0.978 standard_init_linux.go:219: exec user process caused: exec format error
#11 ERROR: executor failed running [/bin/sh -c rustup install nightly]: exit code: 1
------
 > [builder 2/6] RUN rustup install nightly:
------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c rustup install nightly]: exit code: 1
pi@rpie:~/Desktop/nimiq/albatross-lab/core-rs-albatross $ uname -a
Linux rpie 5.10.17-v7l+ #1403 SMP Mon Feb 22 11:33:35 GMT 2021 armv7l GNU/Linux
pi@rpie:~/Desktop/nimiq/albatross-lab/core-rs-albatross $ 
```

Typing `lscpu` stats in same Raspberry Pi model but 32-bit OS:
```
pi@rpie:~/Desktop/nimiq/albatross-lab/core-rs-albatross $ lscpu
Architecture:        armv7l
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               3
Model name:          Cortex-A72
Stepping:            r0p3
CPU max MHz:         1500.0000
CPU min MHz:         600.0000
BogoMIPS:            270.00
Flags:               half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32

```

However, most docker images for Rust particularly the one I used in build_aarch64.Dockerfile only compatible if it runs in 64-bit operating system. [Raspberry Pi OS 64-bit is still in beta build](https://www.raspberrypi.org/forums/viewtopic.php?f=117&t=275370) thus it is not included in official website or in Raspberry Pi Imager.

image

image


```
...
#11 [builder 2/6] RUN rustup install nightly
#11 sha256:1d2abcc85cbf2e06414072232090b01a1eb0a8cb9296628c993fdb1a1fe356a6
#11 CACHED

#12 [builder 3/6] RUN apt-get update && apt-get install -y libssl-dev pkg-config
#12 sha256:1d0a94f8850393fbd330360e2c664ea4a84321c49155b208e85c791f1054ab98
#12 CACHED

#14 [builder 4/6] COPY . /build
#14 sha256:009880c1598005d2f10db4589f1d4ec7c81e134d925a3b08df83708410c292f2
#14 DONE 3.4s

#15 [builder 5/6] WORKDIR /build
#15 sha256:58e0983a49cf78a3e629b4c1f9ed08167ae220079708a1218a4b02b947895654
#15 DONE 0.2s

#16 [builder 6/6] RUN   --mount=type=cache,target=/build/target   --mount=type=cache,target=/usr/local/cargo/registry   --mount=type=cache,target=/usr/local/cargo/git   cargo +nightly build --bin nimiq-client && cp /build/target/debug/nimiq-client /build/
#16 sha256:600c0a322d840cb87e6296c1813dca03f36e317d29d006d36fab8722f0a2f1bf
#16 1.803     Updating crates.io index
#16 49.36     Updating git repository `https://github.com/nimiq/merkle-mountain-range`
#16 49.77     Updating git repository `https://github.com/nimiq/jsonrpc.git`
#16 50.87     Updating git repository `https://github.com/jgraef/warp.git`
#16 53.53  Downloading crates ...
#16 55.60   Downloaded libp2p-dns v0.27.0
#16 55.60   Downloaded lock_api v0.4.4
#16 55.62   Downloaded unsigned-varint v0.6.0
#16 55.63   Downloaded libp2p-yamux v0.30.1
#16 55.66   Downloaded libp2p-gossipsub v0.27.0
#16 55.69   Downloaded openssl v0.10.34
#16 55.72   Downloaded libp2p-swarm v0.27.2
#16 55.74   Downloaded libp2p-request-response v0.9.1
#16 55.74   Downloaded libp2p-plaintext v0.27.1
#16 55.75   Downloaded if-watch v0.1.8
...
```
The installation went ok.

Typing `lscpu` stats in same Raspberry Pi model but 64-bit OS:
```
pi@raspberrypi:~/core-rs-albatross $ lscpu
Architecture:        aarch64
Byte Order:          Little Endian
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
Vendor ID:           ARM
Model:               3
Model name:          Cortex-A72
Stepping:            r0p3
CPU max MHz:         1500.0000
CPU min MHz:         600.0000
BogoMIPS:            108.00
Flags:               fp asimd evtstrm crc32 cpuid
```

image


Also in Raspberry Pi OS 32-bit, I've tried with changing `RUST_IMAGE=arm64v8` to `RUST_IMAGE=arm32v7` and received this:

```
#16 1337.9   thread 'main' panicked at 'Failed to find the protoc binary. The PROTOC environment variable is not set, there is no bundled protoc for this platform, and protoc is not in the PATH', /usr/local/cargo/registry/src/github.com-1285ae84e5963aae/prost-build-0.7.0/build.rs:100:10
#16 1337.9   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
#16 1337.9 warning: build failed, waiting for other jobs to finish...
#16 1348.6 error: build failed
#16 ERROR: executor failed running [/bin/sh -c cargo +nightly build --bin nimiq-client && cp /build/target/debug/nimiq-client /build/]: exit code: 101
------
 > [builder 6/6] RUN   --mount=type=cache,target=/build/target   --mount=type=cache,target=/usr/local/cargo/registry   --mount=type=cache,target=/usr/local/cargo/git   cargo +nightly build --bin nimiq-client && cp /build/target/debug/nimiq-client /build/:
------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c cargo +nightly build --bin nimiq-client && cp /build/target/debug/nimiq-client /build/]: exit code: 101
```

It turned out here, some packages could only work in 64bit.
Regards

**Comment by ha842** [comment]
Pull

**Review by ha842 — APPROVED** [review]

**Comment by jeffesquivels** [comment]
Hi @frvfrvr, sorry for long delay. I have tested your Dockerfile and seems to work fine. Can you squash all the commits into only one and rebase against `albatross` (and also file the PR to be merged to `albatross` instead of `lab-test`), please?

After this I will accept the PR and merge it.

Thanks!

**Comment by frvfrvr** [comment]
> Hi @frvfrvr, sorry for long delay. I have tested your Dockerfile and seems to work fine. Can you squash all the commits into only one and rebase against `albatross` (and also file the PR to be merged to `albatross` instead of `lab-test`), please?
> 
> After this I will accept the PR and merge it.
> 
> Thanks!

Yes sir, I have filed a PR #291 - looking forward for merging. Thank you

**Comment by jeffesquivels** [comment]
Superseded by #291
